src/module.c

```diff
@@ -12754,27 +12843,94 @@ int moduleConfigApplyConfig(list *module_configs, const char **err, const char *
+        const char *delim; /* Locate the '|' delimiter if present */
+        moduleVerifyUnprefixedName(name, &delim);
+        cname = sdsnew(name);
+        if (delim) { /* Handle "<NAME>|<ALIAS>" format */
+            sdssubstr(cname, 0, delim - name);
+            alias = sdsnew(delim + 1);
+        }
+    } else {
+        /* Add the module name prefix */
+        cname = sdscatfmt(sdsempty(), "%s.%s", module->name, name);
+    }
+    
     ModuleConfig *new_config = zmalloc(sizeof(ModuleConfig));
-    new_config->name = sdsnew(name);
+    new_config->unprefixedFlag = flags & REDISMODULE_CONFIG_UNPREFIXED;
+    new_config->name = cname;
+    new_config->alias = alias;
     new_config->apply_fn = apply_fn;
     new_config->privdata = privdata;
     new_config->module = module;
     return new_config;
 }
 
+/* Verify the configuration name and check for duplicates.
+ * 
+ * - If the configuration is flagged as unprefixed, it checks for duplicate
+ *   names and optional aliases in the format <NAME>|<ALIAS>.
+ * - If the configuration is prefixed, it ensures the name is unique with
+ *   the module name prepended (<MODULE_NAME>.<NAME>).
+ */
 int moduleConfigValidityCheck(RedisModule *module, const char *name, unsigned int flags, configType type) {
     if (!module->onload) {
         errno = EBUSY;
         return REDISMODULE_ERR;
     }
-    if (moduleVerifyConfigFlags(flags, type) || moduleVerifyResourceName(name)) {
+    if (moduleVerifyConfigFlags(flags, type)) {
         errno = EINVAL;
         return REDISMODULE_ERR;
     }
-    if (isModuleConfigNameRegistered(module, name)) {
-        serverLog(LL_WARNING, "Configuration by the name: %s already registered", name);
+    
+    int isdup = 0;    
+    if (flags & REDISMODULE_CONFIG_UNPREFIXED) {
+        const char *delim = NULL; /* Pointer to the '|' delimiter in <NAME>|<ALIAS> */
+        if (moduleVerifyUnprefixedName(name, &delim)){
+            errno = EINVAL;
+            return REDISMODULE_ERR;
+        }
+        
+        if (delim) { 
+            /* Temporary split the name and alias string for the check */
+            sds _name = sdsnew(name);            
+            sdssubstr(_name, 0, delim - name);
+            const char *alias = delim + 1;
+            isdup = isModuleConfigNameRegistered(module, _name) ||
+                    isModuleConfigNameRegistered(module, alias);
+            sdsfree(_name);
+        } else {
+            isdup = isModuleConfigNameRegistered(module, name);
+        }
+    } else {
+
+        if (moduleVerifyResourceName(name)) {
+            errno = EINVAL;
+            return REDISMODULE_ERR;
+        }
+
+        sds fullname = sdscatfmt(sdsempty(), "%s.%s", module->name, name);
+        isdup = isModuleConfigNameRegistered(module, fullname);
+        sdsfree(fullname);
+    }
+    
+    if (isdup) {
+        serverLog(LL_WARNING,
+                  "Configuration by the name: %s already registered", name);
```

Comment:
Hi @moticless,
What is the expected behavior if two different modules register config parameters with the same name?

I tested the following case:
1. Load `moduleconfigs`
2. Create and load a new module `moduleconfigs_dup` which registers a single config parameter `unprefix-bool|unprefix-bool-alias`. Duplicated name and alias of a config parameter registered by `moduleconfigs`
3. Unload `moduleconfigs_dup.so`. 
4. The `unprefix-bool|unprefix-bool-alias` config parameters was removed. But it was also part of `module_configs` which is still loaded.

<details>

<summary>Test loading/unloading module - Details</summary>

```sh
127.0.0.1:6379> module loadex /tmp/moduleconfigs.so
OK
127.0.0.1:6379> config get unprefix-bool*
1) "unprefix-bool"
2) "yes"
3) "unprefix-bool-alias"
4) "yes"
127.0.0.1:6379> module loadex /tmp/moduleconfigs_dup.so
OK
127.0.0.1:6379> config get unprefix-bool*
1) "unprefix-bool"
2) "yes"
3) "unprefix-bool-alias"
4) "yes"
127.0.0.1:6379> module unload moduleconfigs_dup
OK
127.0.0.1:6379> config get unprefix-bool*
(empty array)
```


<details>

Reply:
This is a good point, I believe the right behavior is to fail such registration if the configuration is already exists.

Reply:
Pls consider my new commit. I followed @MeirShpilraien's advice.